source/pkg/apis/sources/v1alpha1/rabbitmq_types.go

```diff
@@ -0,0 +1,169 @@
+	"knative.dev/pkg/kmeta"
+	"knative.dev/pkg/webhook/resourcesemantics"
+)
+
+// +genclient
+// +genreconciler
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// RabbitmqSource is the Schema for the rabbitmqsources API.
+// +k8s:openapi-gen=true
+type RabbitmqSource struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   RabbitmqSourceSpec   `json:"spec,omitempty"`
+	Status RabbitmqSourceStatus `json:"status,omitempty"`
+}
+
+var _ runtime.Object = (*RabbitmqSource)(nil)
+var _ resourcesemantics.GenericCRD = (*RabbitmqSource)(nil)
+var _ kmeta.OwnerRefable = (*RabbitmqSource)(nil)
+var _ apis.Defaultable = (*RabbitmqSource)(nil)
+var _ apis.Validatable = (*RabbitmqSource)(nil)
+
+type RabbitmqChannelConfigSpec struct {
+	// Channel Prefetch count
+	// +optional
+	PrefetchCount int  `json:"prefetch_count,omitempty"`
+	// Channel Qos global property
+	// +optional
+	GlobalQos     bool `json:"global_qos,omitempty"`
+}
+
+type RabbitmqSourceExchangeConfigSpec struct {
+	// Name of the exchange
+	// +optional
+	Name        string `json:"name,omitempty"`
+	// Type of exchange e.g. direct, topic, headers, fanout
+	// +required
+	TypeOf      string `json:"type,omitempty"`
+	// Exchange is Durable or not
+	// +optional
+	Durable     bool   `json:"durable,omitempty"`
+	// Exchange can be AutoDeleted or not
+	// +optional
+	AutoDeleted bool   `json:"auto_detected,omitempty"`
+	// Declare exchange as internal or not.
+	// Exchanges declared as `internal` do not accept accept publishings.
+	// +optional
+	Internal    bool   `json:"internal,omitempty"`
+	// Exchange NoWait property. When set to true,
+	// declare without waiting for a confirmation from the server.
+	// +optional
+	NoWait      bool   `json:"nowait,omitempty"`
+}
+
+type RabbitmqSourceQueueConfigSpec struct {
+	// Name of the queue to bind to.
+	// Queue name may be empty in which case the server will generate a unique name.
+	// +optional
+	Name             string `json:"name,omitempty"`
+	// Routing key of the messages to be received.
+	// Multiple routing keys can be specified separated by commas. e.g. key1,key2
+	// +optional
+	RoutingKey       string `json:"routing_key,omitempty"`
+	// Queue is Durable or not
+	// +optional
+	Durable          bool   `json:"durable,omitempty"`
+	// Queue can be AutoDeleted or not
+	// +optional
+	DeleteWhenUnused bool   `json:"delete_when_unused,omitempty"`
+	// Queue is exclusive or not.
+	// +optional
+	Exclusive        bool   `json:"exclusive,omitempty"`
+	// Queue NoWait property. When set to true,
+	// the queue will assume to be declared on the server.
+	// +optional
+	NoWait           bool   `json:"nowait,omitempty"`
+}
+
+type RabbitmqSourceSpec struct {
+	// Brokers are the Rabbitmq servers the consumer will connect to.
+	// +required
+	Brokers string `json:"brokers"`
+	// Topic topic to consume messages from
+	// +required
+	Topic   string `json:"topic,omitempty"`
```

Comment:
why is this here? aren't message consumed from the Queue that is bound to the Exchange?